phlib/hndlinfo.c

```diff
@@ -836,95 +836,113 @@ _Callback_ PPH_STRING PhStdGetClientIdName(
     _In_ PCLIENT_ID ClientId
     )
 {
-    static PH_QUEUED_LOCK cachedProcessesLock = PH_QUEUED_LOCK_INIT;
-    static PVOID processes = NULL;
-    static ULONG64 lastProcessesTickCount = 0;
-
-    PPH_STRING name;
-    ULONG64 tickCount;
-    PSYSTEM_PROCESS_INFORMATION processInfo;
+    return PhStdGetClientIdNameEx(ClientId, NULL);
+}
 
-    // Get a new process list only if 2 seconds have passed since the last update.
-    tickCount = NtGetTickCount64();
+PPH_STRING PhStdGetClientIdNameEx(
+    _In_ PCLIENT_ID ClientId,
+    _In_opt_ PPH_STRING ProcessName
+    )
+{
+    PPH_STRING result;
+    PPH_STRING processName = NULL;
+    PPH_STRING threadName = NULL;
+    PH_STRINGREF processNameRef;
+    PH_STRINGREF threadNameRef;
+    HANDLE hProcess, hThread;
+    ULONG isProcessTerminated = FALSE;
+    ULONG isThreadTerminated = FALSE;
 
-    if (tickCount - lastProcessesTickCount >= 2000)
+    if (ProcessName)
     {
-        PhAcquireQueuedLockExclusive(&cachedProcessesLock);
-
-        // Re-check the tick count.
-        if (tickCount - lastProcessesTickCount >= 2000)
+        // Use the supplied name
+        processNameRef.Length = ProcessName->Length;
+        processNameRef.Buffer = ProcessName->Buffer;
+    }
+    else
+    {
+        // Determine the name of the process ourselves
+        if (NT_SUCCESS(PhGetProcessImageFileNameById(
+            ClientId->UniqueProcess,
+            NULL,
+            &processName
+            )))
         {
-            if (processes)
-            {
-                PhFree(processes);
-                processes = NULL;
-            }
-
-            if (!NT_SUCCESS(PhEnumProcesses(&processes)))
-            {
-                PhReleaseQueuedLockExclusive(&cachedProcessesLock);
-                return PhCreateString(L"(Error querying processes)");
-            }
-
-            lastProcessesTickCount = tickCount;
+            processNameRef.Length = processName->Length;
+            processNameRef.Buffer = processName->Buffer;
+        }
+        else
+        {
+            PhInitializeStringRef(&processNameRef, L"Unknown process");
         }
-
-        PhReleaseQueuedLockExclusive(&cachedProcessesLock);
     }
 
-    // Get a lock on the process list and get a name for the client ID.
+    // Check if the process is alive, but only if we didn't get its name
+    if (!ProcessName && NT_SUCCESS(PhOpenProcess(
+        &hProcess,
+        SYNCHRONIZE,
+        ClientId->UniqueProcess
+        )))
+    {
+        LARGE_INTEGER timeout = { 0 };
 
-    PhAcquireQueuedLockShared(&cachedProcessesLock);
+        // Waiting with zero timout checks for termination
```

Comment:
Shouldn't this be calling PhGetProcessBasicInformation(hProcess)? ExitStatus will always return STATUS_PENDING when the process hasn't terminated.

Reply:
I thought about it, but what if a processes exists with STATUS_PENDING?

Reply:
It's mentioned in the documentation for GetExitCodeProcess that programs shouldn't ever return STATUS_PENDING since it can cause infinate loops but we don't use loops so it shouldn't be an issue?

If you think it should use NtWaitForSingleObject then it can remain there. I just wanted to double check if that was intended?

Reply:
I don't mind doing either way. I see NtWaitForSingleObject slightly more reliable when it comes to false negatives, but it requires opening the target for SYNCHRONIZE rather than PROCESS_QUERY_LIMITED_INFORMATION. I don't know which one is better, to be honest.

Reply:
I guess we'll find out which way works better eventually 😋